File: mobile/src/bindings/category.ts
import { invoke } from "@tauri-apps/api/core";
import type { CategoryView } from "./types";

export const category = {
  async add(name: string): Promise<void> {
    await invoke("category", { action: "Add", name });
  },

  async remove(name: string): Promise<void> {
    await invoke("category", { action: "Remove", name });
  },

  async view(name: string): Promise<CategoryView> {
    return await invoke("category", { action: "View", name });
  },

  async list(): Promise<string[]> {
    return await invoke("category", { action: "List" });
  },

  async assign(ingredientName: string, categories: string[]): Promise<void> {
    await invoke("category", {
      action: "Assign",
      ingredientname: ingredientName,
      categories,
    });
  },

  async unassign(ingredientName: string, categories: string[]): Promise<void> {
    await invoke("category", {
      action: "Unassign",
      ingredientname: ingredientName,
      categories,
    });
  },
};

File: mobile/src-tauri/src/commands/category.rs
use foody_core::app::App;
use serde::Deserialize;
use tauri::State;

use crate::commands::serialize_json;

#[derive(Deserialize)]
#[serde(tag = "action")]
pub enum CategoryCommand {
    Add {
        name: String,
    },
    Remove {
        name: String,
    },
    View {
        name: String,
    },
    List,
    Assign {
        ingredientname: String,
        categories: Vec<String>,
    },
    Unassign {
        ingredientname: String,
        categories: Vec<String>,
    },
}

#[tauri::command]
pub async fn category(
    state: State<'_, App>,
    command: CategoryCommand,
) -> Result<serde_json::Value, String> {
    match command {
        CategoryCommand::Add { name } => serialize_json(state.add_category(&name)).await,
        CategoryCommand::Remove { name } => serialize_json(state.remove_category(&name)).await,
        CategoryCommand::View { name } => serialize_json(state.view_category(&name)).await,
        CategoryCommand::List => serialize_json(state.list_categories()).await,
        CategoryCommand::Assign {
            ingredientname,
            categories,
        } => serialize_json(state.assign_categories(&ingredientname, &categories)).await,
        CategoryCommand::Unassign {
            ingredientname,
            categories,
        } => serialize_json(state.unassign_categories(&ingredientname, &categories)).await,
    }
}
